plugins/chat/app/services/chat/add_users_to_channel.rb

```diff
@@ -60,17 +57,36 @@ def can_add_users_to_channel(guardian:, channel:, **)
     end
 
     def fetch_users(contract:, channel:, **)
-      ::User.where(
-        "username IN (?) AND id NOT IN (?)",
-        [*contract.usernames],
-        channel.chatable.direct_message_users.select(:user_id),
-      ).to_a
+      ::User
+        .joins(:user_option)
+        .left_outer_joins(:groups)
+        .where(user_options: { chat_enabled: true })
+        .where(username: [*contract.usernames])
+        .or(
+          User
+            .joins(:user_option)
+            .left_outer_joins(:groups)
+            .where(user_options: { chat_enabled: true })
+            .where(groups: { name: contract.groups })
+            .where.not(group_users: { user_id: nil }),
+        )
+        .where.not(id: channel.chatable.direct_message_users.select(:user_id))
+        .distinct
+        .to_a
```

Comment:
Could you make a query object form this query and the same one you use in create?

Something like this:

```ruby
UsersFromUsernamesAndGroups.call(usernames:  [], groups: [])
```